File: documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/RemoveLocationMessage.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.documentapi.messagebus.protocol;

import com.yahoo.document.*;
import com.yahoo.document.select.BucketSelector;
import java.util.Set;

/**
 * Message (VDS only) to remove an entire location for users using n= or g= schemes.
 * We use a document selection so the user can specify a subset of those documents to be deleted
 * if they wish.
 */
public class RemoveLocationMessage extends DocumentMessage {
    String documentSelection;
    BucketId bucketId;

    public RemoveLocationMessage(String documentSelection) {
        try {
            this.documentSelection = documentSelection;
            BucketSelector bucketSel = new BucketSelector(new BucketIdFactory());
            Set<BucketId> rawBuckets = bucketSel.getBucketList(documentSelection);
            if (rawBuckets == null || rawBuckets.size() != 1) {
                throw new IllegalArgumentException("Document selection for remove location must map to a single location (user or group)");
            } else {
                // There can only be one.
                for (BucketId id : rawBuckets) {
                    bucketId = id;
                }
            }
        } catch (com.yahoo.document.select.parser.ParseException p) {
            throw new IllegalArgumentException(p);
        }
    }

    public String getDocumentSelection() {
        return documentSelection;
    }

    @Override
    public DocumentReply createReply() {
        return new DocumentReply(DocumentProtocol.REPLY_REMOVELOCATION);
    }

    @Override
    public int getType() {
        return DocumentProtocol.MESSAGE_REMOVELOCATION;
    }

    public BucketId getBucketId() {
        return bucketId;
    }
}

File: documentapi/src/vespa/documentapi/messagebus/messages/removelocationmessage.cpp
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "removelocationmessage.h"
#include <vespa/documentapi/messagebus/documentprotocol.h>
#include <vespa/document/select/parser.h>
#include <vespa/document/bucket/bucketselector.h>
#include <vespa/vespalib/util/exceptions.h>

namespace documentapi {

RemoveLocationMessage::RemoveLocationMessage(
        const document::BucketIdFactory& factory,
        document::select::Parser& parser,
        const string& documentSelection)
    : _documentSelection(documentSelection),
      _bucketId(),
      _bucketSpace()
{
    document::BucketSelector bucketSel(factory);
    std::unique_ptr<document::BucketSelector::BucketVector> exprResult(
            bucketSel.select(*parser.parse(documentSelection)));

    if (exprResult.get() && exprResult->size() == 1) {
        _bucketId = (*exprResult)[0];
    } else {
        throw vespalib::IllegalArgumentException(
                "Document selection doesn't map to a single bucket!",
                VESPA_STRLOC);
    }
}

RemoveLocationMessage::~RemoveLocationMessage() {
}

DocumentReply::UP
RemoveLocationMessage::doCreateReply() const {
    return DocumentReply::UP(
            new DocumentReply(DocumentProtocol::REPLY_REMOVELOCATION));
}

uint32_t
RemoveLocationMessage::getType() const {
    return DocumentProtocol::MESSAGE_REMOVELOCATION;
}

}
